src/AudioInterface.cpp

```diff
@@ -497,44 +507,39 @@ void AudioInterface::fromSampleToBitConversion
  const AudioInterface::audioBitResolutionT targetBitResolution)
 {
     int8_t tmp_8;
-    uint8_t tmp_u8; // unsigned to quantize the remainder in 24bits
     int16_t tmp_16;
-    sample_t tmp_sample;
-    sample_t tmp_sample16;
-    sample_t tmp_sample8;
+    double tmp_sample;
     switch (targetBitResolution)
     {
     case BIT8 :
         // 8bit integer between -128 to 127
-        tmp_sample = floor( (*input) * 128.0 ); // 2^7 = 128.0
+        tmp_sample = std::max(-127.0, std::min(127.0, std::round( (*input) * 127.0 ))); // 2^7 = 128
         tmp_8 = static_cast<int8_t>(tmp_sample);
         std::memcpy(output, &tmp_8, 1); // 8bits = 1 bytes
         break;
     case BIT16 :
         // 16bit integer between -32768 to 32767
-        tmp_sample = floor( (*input) * 32768.0 ); // 2^15 = 32768.0
+        // tmp_sample = floor( (*input) * 32768.0 ); // 2^15 = 32768.0
+        tmp_sample = std::max(-32767.0, std::min(32767.0, std::round( (*input) * 32767.0 ))); // 2^15 = 32768
         tmp_16 = static_cast<int16_t>(tmp_sample);
-        std::memcpy(output, &tmp_16, 2); // 16bits = 2 bytes
-        break;
-    case BIT24 :
-        // To convert to 24 bits, we first quantize the number to 16bit
-        tmp_sample = (*input) * 32768.0; // 2^15 = 32768.0
-        tmp_sample16 = floor(tmp_sample);
-        tmp_16 = static_cast<int16_t>(tmp_sample16);
-
-        // Then we compute the remainder error, and quantize that part into an 8bit number
-        // Note that this remainder is always positive, so we use an unsigned integer
-        tmp_sample8 = floor ((tmp_sample - tmp_sample16)  //this is a positive number, between 0.0-1.0
-                             * 256.0);
-        tmp_u8 = static_cast<uint8_t>(tmp_sample8);
-
-        // Finally, we copy the 16bit number in the first 2 bytes,
-        // and the 8bit number in the third bite
-        std::memcpy(output, &tmp_16, 2); // 16bits = 2 bytes
-        std::memcpy(output+2, &tmp_u8, 1); // 8bits = 1 bytes
+        std::memcpy(output, &tmp_16, 2); // 2 bytes output in Little Endian order (LSB -> smallest address)
         break;
+    case BIT24 : {
+        // tmp_sample = (*input) * 32768.0; // 2^15 = 32768.0
+        tmp_sample = std::max(-8388607.0, std::min(8388607.0, std::round( (*input) * 8388607.0 ))); // 2^23 = 8388608
+        union {
+          uint32_t tmp32;
+          char tmp32c[4];
+        } tmp32u;
+        tmp32u.tmp32 = tmp_sample;
+        int offset = ( is_big_endian() ? 1 : 0 );
+        std::memcpy(output, &(tmp32u.tmp32c[offset]), 3); // 24bits = 3 bytes
```

Comment:
The 24-bit implementation looks pretty weird to me - your new 24-bit implementation looks like it's incompatible with the previous version - is this breaking change intentional? it looks like it would mean that older clients are incompatible with newer clients when using 24-bit mode.

Reply:
I susppect 24-bit mode is un-used in practice, but I think it's better to explicit about these things.

Reply:
and presumably you should also be updating this block https://github.com/jacktrip/jacktrip/blob/main/src/AudioInterface.cpp#L516-L526 in the same PR?

Reply:
The previous 24-bit implementation is [documented as not working](https://github.com/jacktrip/jacktrip/blob/9f596d3778b6614650d9253e201ceeb67abed938/src/AudioInterface.cpp#L442).  I therefore did not test in detail how my implementation would behave differently in all 24-bit cases, but rather wrote what I thought was a good approach for handling the conversion in an endian-independent way, intending to put out the same stream of byes in the little-endian non-clipped case.   (I did single-step it in my testing to see that it was doing what I wanted.)  As a result, I expected my 24-bit case to behave differently on big-endian machines, but the same across big/little endian platforms, which I felt was more important.   I didn't worry about breaking cases involving a new or old big-endian or little-endian clients or servers in 24-bit mode, e.g., because the -b 24 feature is documented as not working, and I've never encountered a server in that mode.  While I consider what I wrote to be a move in the right direction, I agree it is not a complete solution for the 24-bit case, since I didn't rewrite the other direction as well.

Since I "touched" the 24-bit case, it is reasonable to ask me to fix it in both directions in this PR, but I think that should happen in a separate PR under current recommendations.  I therefore conclude that I should back out of my recommended change to the 24-bit case in this PR so that both directions can be addressed in a PR aimed at fixing 24-bit support (which, as you note, nobody seems to use).

Reply:
Yes, I agree the 24-bit support should be addressed separately - my PR got delayed because I didn't have time to get to the bottom of the 24-bit handling & it would be really good to get a 16-bit fix merged and in the hands of users. 

To be honest, I would suggest as a next step removing / disabling 24-bit support in JackTrip, pending a re-write.